### ISSUE TITLE
`r.turing`

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5948]
`input.turing(machine)` evaluates the turing machine specified by `machine` (see below) on the initial tape given by `input` (an array of integers). It then returns an array representing the prefix of the tape that has been accessed by the machine. If the input is rejected, an exception is thrown.

`machine` is an object as follows:
```js
{
  blank_symbol: int,
  transitions: [
    {
      from_state: int,
      read_symbol: int,
      to_state: int,
      write_symbol: int,
      move_head: "L"|"R"
    },
    ...
  ],
  initial_state: int,
  final_states: [int, ...]
}
```

--

The implementation is simply as a query rewrite in terms of other ReQL operations.

Some helper functions:
```js
var initialState = { tape: input, state: machine.initial_state, head_position: 0 };

// Reads the current symbol from the tape.
var readSymbol = function(state) {
    return state('tape').nth(state('head_position')).default(machine.blank_symbol);
  };

// Tests whether the given transition applies to the current state.
var checkRule = function(state, rule) {
    return state('state').eq(rule('from_state')).and(readSymbol(state).eq(rule('read_symbol')));
  };

// Selects an applicable transition.
var selectRule = function(state) {
    return r.expr(machine.transitions)
      .concatMap(function(rule) {
        return r.branch(checkRule(state, rule), [rule], []);
      })
      .nth(0)
      .default(r.error("No applicable rule. Input rejected."));
  };

// Writes `newSymbol` to the current head position and returns the new tape.
var writeTape = function(state, newSymbol) {
  // Note that head_position is never larger than the current tape length
  var mustAppend = state('tape').count().eq(state('head_position'));
  return r.branch(
    mustAppend,
    state('tape').append(newSymbol),
    state('tape').changeAt(state('head_position'), newSymbol));
};

// Applies the given transition to the given state, and returns a new state.
var applyRule = function(state, rule) {
  return {
    tape: writeTape(state, rule('write_symbol')),
    state: rule('to_state'),
    head_position: state('head_position').add(r.branch(rule('move_head').eq("L"), -1, 1))
  };
};
```

The actual query:
```js
r.range()
 .fold(
   initialState,
   function(state, _ignored) {
     // Select a transition and compute the new state by applying it.
     var selectedRule = selectRule(state);
     return applyRule(state, selectedRule);
   },
   {
     emit: function(_ignored, _ignored2, state) {
       // Check if we have reached a final state. If yes, accept the input by emitting the current tape.
       return r.branch(
         r.expr(machine.final_states).contains(state('state')),
         [state('tape')],
         []);
     }
   }
 )
 .limit(1) // Stop as soon as we get to an accepted state
```

**Note:** This is not an actual serious proposal. I just thought it was cool that you could do this in ReQL.

**Warning**: Don't run this on a production server. It's easy to end up with a query that never terminates, and I'm not sure if those queries can be interrupted.

**Comment by danielmewes** [comment]
Here's an example machine and input that you can paste and run in the Data Explorer.
It just writes the symbol 1 onto the tape, transitions to state 1 and terminates.

```js
var machine = {
  blank_symbol: 0,
  transitions: [
    {
      from_state: 0,
      read_symbol: 0,
      to_state: 1,
      write_symbol: 1,
      move_head: "R"
    }
  ],
  initial_state: 0,
  final_states: [1]
};

var input = [];

var initialState = { tape: input, state: machine.initial_state, head_position: 0 };

// Reads the current symbol from the tape.
var readSymbol = function(state) {
    return state('tape').nth(state('head_position')).default(machine.blank_symbol);
  };

// Tests whether the given transition applies to the current state.
var checkRule = function(state, rule) {
    return state('state').eq(rule('from_state')).and(readSymbol(state).eq(rule('read_symbol')));
  };

// Selects an applicable transition.
var selectRule = function(state) {
    return r.expr(machine.transitions)
      .concatMap(function(rule) {
        return r.branch(checkRule(state, rule), [rule], []);
      })
      .nth(0)
      .default(r.error("No applicable rule. Input rejected."));
  };

// Writes `newSymbol` to the current head position and returns the new tape.
var writeTape = function(state, newSymbol) {
  // Note that head_position is never larger than the current tape length
  var mustAppend = state('tape').count().eq(state('head_position'));
  return r.branch(
    mustAppend,
    state('tape').append(newSymbol),
    state('tape').changeAt(state('head_position'), newSymbol));
};

// Applies the given transition to the given state, and returns a new state.
var applyRule = function(state, rule) {
  return {
    tape: writeTape(state, rule('write_symbol')),
    state: rule('to_state'),
    head_position: state('head_position').add(r.branch(rule('move_head').eq("L"), -1, 1))
  };
};


r.range()
 .fold(
   initialState,
   function(state, _ignored) {
     // Select a transition and compute the new state by applying it.
     var selectedRule = selectRule(state);
     return applyRule(state, selectedRule);
   },
   {
     emit: function(_ignored, _ignored2, state) {
       // Check if we have reached a final state. If yes, accept the input by emitting the current tape.
       return r.branch(
         r.expr(machine.final_states).contains(state('state')),
         [state('tape')],
         []);
     }
   }
 )
 .limit(1) // Stop as soon as we get to an accepted state
```

**Comment by danielmewes** [comment]
To clarify: What this shows is that you can run any turing machine in ReQL right now, using only functionality included in RethinkDB 2.3. This proves that ReQL is a turing-complete language.
